os/kernel/task/task_event_wait.c

```diff
@@ -0,0 +1,86 @@
+/****************************************************************************
+ *
+ * Copyright 2020 Samsung Electronics All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied. See the License for the specific
+ * language governing permissions and limitations under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <assert.h>
+#include <arch/irq.h>
+#include <tinyara/arch.h>
+#include <tinyara/sched.h>
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+static void update_event_queue(struct tcb_s *tcb)
+{
+	if (tcb->queueing == true) {
+		tcb->queueing = false;
+		tcb->event_data = tcb->pending_event_data;
+	} else {
+		tcb->event_data = NO_PENDING_DATA;
+	}	
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: task_event_wait
+ *
+ * Description:
+ *   This function is for waiting kernel event.
+ *   If a task calls this function, it will be blocked until receiving event.
+ *
+ * Inputs: None
+ *
+ * Return Value:
+ *   On success, passed data will be returned.
+ *   On failure, 0 will be returned.
+ *
+ ****************************************************************************/
+
+void *task_event_wait(void)
+{
+	void *ret;
+	int saved_state;
+	struct tcb_s *tcb = sched_self();
+	DEBUGASSERT(tcb);
+
+	saved_state = irqsave();
+	if (tcb->event_data != NO_PENDING_DATA) {
+		/* If there is already received event, return saved event data. */
+		ret = tcb->event_data;
+		/* If there is pending event, update the event data array. */
+		update_event_queue(tcb);
+		irqrestore(saved_state);
+		return ret;
+	}
+
+	up_block_task(tcb, TSTATE_WAIT_UNBLOCK);
+
+	ret = tcb->event_data;
+	/* If there is pending event, update the event data array. */
+	update_event_queue(tcb);
```

Comment:
This scope will be reached if tcb->event_data is NO_PENDING_DATA, then no update needed. I think.

Reply:
Thanks for the comment.